internal/stacks/stackruntime/validate_test.go

```diff
@@ -356,3 +356,68 @@ Terraform uses references to decide a suitable order for performing operations,
 		t.Errorf("wrong diagnostics\n%s", diff)
 	}
 }
+
+func TestValidate_impliedProviderTypes(t *testing.T) {
+
+	tcs := []struct {
+		directory string
+		providers map[addrs.Provider]providers.Factory
+		wantDiags func() tfdiags.Diagnostics
+	}{
+		{
+			directory: "with-hashicorp-provider",
+			providers: map[addrs.Provider]providers.Factory{
+				addrs.NewDefaultProvider("testing"): func() (providers.Interface, error) {
+					return stacks_testing_provider.NewProvider(), nil
+				},
+			},
+		},
+		{
+			directory: "with-non-hashicorp-provider",
+			providers: map[addrs.Provider]providers.Factory{
+				addrs.NewProvider(addrs.DefaultProviderRegistryHost, "other", "testing"): func() (providers.Interface, error) {
+					return stacks_testing_provider.NewProvider(), nil
+				},
+			},
+			wantDiags: func() tfdiags.Diagnostics {
+				var diags tfdiags.Diagnostics
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid provider configuration",
+					Detail: "The provider configuration slot \"testing\" requires a configuration for provider \"registry.terraform.io/hashicorp/testing\", not for provider \"registry.terraform.io/other/testing\"." +
+						"\n\nThe provider type required by the module has been automatically implied by Terraform, explicitly setting the provider type within the modules required_providers block may resolve this issue.",
+					Subject: &hcl.Range{
+						Filename: mainBundleSourceAddrStr("legacy-module/with-non-hashicorp-provider/with-non-hashicorp-provider.tfstack.hcl"),
+						Start:    hcl.Pos{Line: 21, Column: 15, Byte: 447},
+						End:      hcl.Pos{Line: 21, Column: 39, Byte: 471},
+					},
+				})
+				return diags
+			},
+		},
+		// TODO: Add a test for deeply nested modules, the middle module still needs a required provider block.
```

Comment:
TODOs left

Reply:
Yeah also left over, I changed the way the errors are being deduced so I don't think we need specific tests for these anymore.